go/service/simplefs.go

```diff
@@ -12,6 +14,58 @@ import (
 	"golang.org/x/net/context"
 )
 
+// transporterWaiter exposes a get() method to get a transporter, and waits for
+// it if it's not available yet. Additionally, it also tries to have get()
+// calls from different goroutines return in the same order they come in. This
+// is done by assigning a number to each call, and process them one-by-one.
+// Only the current processing goroutine actively checks and waits for the
+// transporter. Other calls won't start until their previous ones have
+// finished.
+type transporterWaiter struct {
+	getter func() rpc.Transporter
+
+	// Just use atomic instead of having cond protect this, to mimize the
+	// chance that multiple calls wait on the same Mutex (which is not
+	// necessarily FIFO).
+	lastWaitingNumber uint64
+
+	cond                *sync.Cond
+	lastProcessedNumber uint64
+}
+
+func newTransporterWaiter(
+	tranporterGetter func() (transporter rpc.Transporter),
+) *transporterWaiter {
+	return &transporterWaiter{
+		getter:              tranporterGetter,
+		lastWaitingNumber:   0,
+		cond:                sync.NewCond(&sync.Mutex{}),
+		lastProcessedNumber: 0,
+	}
+}
+
+const waitForTransporterInterval = time.Second / 4
+
+func (w *transporterWaiter) get() (xp rpc.Transporter) {
+	myNumber := atomic.AddUint64(&w.lastWaitingNumber, 1)
+
+	w.cond.L.Lock()
+	defer w.cond.L.Unlock()
+	for w.lastProcessedNumber != myNumber-1 {
+		w.cond.Wait()
+	}
+
+	xp = w.getter()
+	for xp == nil {
```

Comment:
Shouldn't there be a timeout here, and/or a context `Done()` check?  I feel like we don't want to wait indefinitely, and we for sure shouldn't hang if the user cancels the request when they get tired of waiting.